commons/src/main/java/org/infinispan/commons/api/functional/FunctionalMap.java

```diff
@@ -0,0 +1,634 @@
+       * {@link CloseableIterator} that allows navigating completion
+       * of the operation executions.
+       *
+       * This method can be used to implement operations such as
+       * {@link javax.cache.Cache#removeAll(Set)}.
+       *
+       * DESIGN RATIONALE:
+       * <ul>
+       *    <li>It makes sense to expose global operation like this
+       *    instead of forcing users to iterate over the keys to lookup and call
+       *    get individually since Infinispan can do things more efficiently.
+       *    </li>
+       *    <li>Since this is a write-only operation, no entry attributes can be
+       *    queried, hence the only reasonable thing can be returned is Void.
+       *    </li>
+       *    <li>It still makes sense to return a CloseableIterator<Void> instead of
+       *    CompletableFuture<Void> in case the user wants to do something as each
+       *    operation completes, e.g. keep track of progress.
+       *    </li>
+       * </ul>
+       *
+       * @param keys the keys associated with each of the {@link WriteEntryView}
+       *             passed in the function callbacks
+       * @param f operation that the {@link WriteEntryView} associated with
+       *          one of the keys passed in
+       * @return a {@link CloseableIterator} that can be navigated to follow progress
+       *         as operations get invoked
+       */
+      CloseableIterator<Void> evalMany(Set<? extends K> keys, Consumer<WriteEntryView<V>> f);
+
+      /**
+       * Evaluate a write-only {@link Consumer} operation with the
+       * {@link WriteEntryView} of the value associated with the key, for all
+       * existing keys in functional map, and returns a
+       * {@link CloseableIterator} that allows navigating completion
+       * of the operation executions.
+       *
+       * This method can be used to implement operations such as
+       * {@link javax.cache.Cache#removeAll()}.
+       *
+       * DESIGN RATIONALE:
+       * <ul>
+       *    <li>In the original design, this method would have taken no
+       *    parameters and returned a CloseableIterator<WriteEntryView<V>>,
+       *    but doing so would have created encapsulation problems, since any
+       *    changes to the entry view would need to be committed, and that
+       *    would happen outside of the interceptor stack scope. To avoid this
+       *    problem, simply pass in a consumer with the desired operation and
+       *    then, the interceptor stack can control how the consumer gets
+       *    executed and make sure it's committed.</li>
+       * </ul>
+       *
+       * @param f operation that the {@link WriteEntryView} associated with
+       *          one of the keys passed in
+       * @return a {@link CloseableIterator} that can be navigated to follow progress
+       *         as operations get invoked
+       */
+      CloseableIterator<Void> evalAll(Consumer<WriteEntryView<V>> f);
```

Comment:
Re: Iterator<Void>

^ The return is not very useful but it allows you to keep track of the progress, even if it's with little information, and it allows you to know when the end has been reached.

Re: Idempotency.

^ What about we use a similar approach to what I temporarily did with the value matcher stuff in `@SerializeWith`? The value matcher is essentially a way to provide guarantees around the results of functions when retried. Here we are trying to find a way to distinguish between normal and retried executions. The only problem about adding it to @SerializeWith is that you need to rely on stateless functions (since you don't control instantiation), but that way, we avoid polluting the end user API.